package jetbrains.mps.baseLanguage.unitTest.execution.server;

/*Generated by MPS */

import java.io.PrintStream;
import org.apache.log4j.BasicConfigurator;
import org.jetbrains.annotations.NotNull;
import org.junit.runner.notification.RunListener;
import org.junit.runner.Request;
import java.io.IOException;
import jetbrains.mps.tool.environment.Environment;
import jetbrains.mps.tool.environment.EnvironmentContainer;

public class DefaultTestExecutor extends AbstractTestExecutor {
  public static final int EXIT_CODE_FOR_EXCEPTION = -12345;

  private final CommandOutputStream myOutStream;
  private final CommandOutputStream myErrStream;

  public DefaultTestExecutor(TestsContributor testsContributor) {
    super(testsContributor);
    myOutStream = new CommandOutputStream(System.out);
    myErrStream = new CommandOutputStream(System.err);
  }

  public void init() {
    System.setOut(new PrintStream(myOutStream));
    System.setErr(new PrintStream(myErrStream));
    BasicConfigurator.configure();
  }

  public void dispose() {
    System.setOut(myOutStream.getOldStream());
    System.setErr(myErrStream.getOldStream());
  }

  @NotNull
  @Override
  protected RunListener createListener(Iterable<Request> iterable) {
    return new DefaultRunListener(myOutStream);
  }

  /**
   * Called when BTestCase is executed
   */
  public static void main(String[] args) throws ClassNotFoundException, IOException {
    DefaultTestExecutor executor = new DefaultTestExecutor(new CommandLineTestsContributor(args));
    try {
      executor.run();
    } catch (Throwable t) {
      executor.processThrowable(t);
    }
    executor.exit();
  }

  protected void run() {
    try {
      init();
      execute();
    } finally {
      dispose();
    }
  }

  @Override
  protected void processThrowable(Throwable t) {
    super.processThrowable(t);
    t.printStackTrace(System.err);
  }

  protected void exit() {
    Environment env = EnvironmentContainer.get();
    if (env != null) {
      env.dispose();
    }
    if (getExecutionError() != null) {
      System.exit(EXIT_CODE_FOR_EXCEPTION);
    } else {
      System.exit(getFailureCount());
    }
  }
}
